### DISCUSSION TITLE
.select() right after .onAuthStateChange() returns empty array

### DISCUSSION THREAD

**Discussion by rishimohan** [supabase/supabase#5391] [Questions] [answered]
I have to fetch data right after user logs in. This is how I'm doing it:

```
const { data: authListener } = sb.auth.onAuthStateChange(
  async (event, session) => {
    if (event == "SIGNED_IN") {
      setSession(session?.user ?? false);
    }
  }
);

useEffect(() => {
  if(session?.id) {
    const { data, error } = await db.select();

   // Data === [], data returns empty
  }
}, [session]);

```

`sb` is the supabase client, `db` is supabase client set with the table name. I have RLS enabled with policy `auth.uid() = user_id`

In above, the `data` returned is an empty array if though there's a matching row. But when I refresh the page the `data` correct logs the row data. I might be missing something, would be great if anyone can help. Thanks.

**Comment by GaryAustin1** [comment, marked as answer]
Try putting the .from part into your useEffect.  I believe that might be where it sets the needed header info with the user jwt.

**Reply by rishimohan** [reply]
Dang! I had spent 3 hours into this and nothing worked, your suggestion did! Thanks Gary a tons!
